### PR TITLE
Blog for NativeJDB - a Java debugger for native images

### PR DESCRIPTION
**Pull request by aavarghese**
** If you are updating a guide, please submit your pull request to the main repository: https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc **

### CODE REVIEW

File: _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+
+https://github.com/nativejdb/[NativeJDB] is an *open-source* tool that supports *remote debugging* of a natively compiled Java code. In principle, this allows debugging from any IDE that supports the Java Platform Debugger Architecture (JPDA), such as IntelliJ, Eclipse or VSCode. Attaching remotely means that the user can start a debugging session on *any O/S*, while the NativeJDB process itself runs in a Linux container. Our challenge is to bridge the gap between JPDA and the GDB, two stacks that don't typically talk to each other! To achieve this we essentially teach the GDB to speak the Java Debug Wire Protocol (JDWP), used by IDEs to communicate with Java debuggers.
```
Comment (maxandersen) [inline _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc:28]: does this mean it only works on debugging linux binaries? no osx nor windows debug hosts ? just want to clarify if that is the implication.

File: _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+
+When writing Quarkus code, developers have the possibility of finding bugs with a regular Java debugger before compiling it natively. NativeJDB is useful when the observed behavior of the native executable is different than when the program is executed on a JVM. It can be used to explore optimizations and differences in libraries that account for behavioral anomalies. NativeJDB is not specific to Quarkus and can be used to debug any natively compiled Java code.
+
+
+[#architecture]
+== Architecture and Implementation
+
+
+NativeJDB is a Java process that wraps the GDB, which itself wraps the native executable, and it acts as a server that understands the JDWP protocol. It is similar to the JDWP agent used to control and debug a running Java process. In that case, the user can attach to such a process and can start debugging it. NativeJDB is used in a similar way: once it is started, the user attaches to it using a Remote Debug configuration from the IDE.
```
Comment (maxandersen) [inline _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc:38]:
```suggestion
NativeJDB is a Java process that wraps GDB, which itself wraps the native executable, and it acts as a server that understands the JDWP protocol. It is similar to the JDWP agent used to control and debug a running Java process. In that case, the user can attach to such a process and can start debugging it. NativeJDB is used in a similar way: once it is started, the user attaches to it using a Remote Debug configuration from the IDE.
```

File: _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+{empty} +
+
+
+image::nativejdbarch_1.png[alt=NativeJDBArchitecture_1]
+
+
+{empty} +
+
+
+Communication between the IDE and the GDB can be both synchronous and asynchronous, including error messages. In some cases, a command from the IDE can result in both a synchronous and an asynchronous response. For example, setting a breakpoint has a synchronous response, but the corresponding breakpoint-hit is an asynchronous event that gets sent from NativeJDB to the IDE. In principle, NativeJDB can be used with any IDE that has a JDWP implementation, but it has mostly been tested and works with IntelliJ. It also does not require any changes or extensions to existing IDEs and works with the *community version* of IntelliJ.
```
Comment (maxandersen) [inline _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc:53]:
```suggestion
Communication between the IDE and the GDB can be both synchronous and asynchronous, including error messages. In some cases, a command from the IDE can result in both a synchronous and an asynchronous response. For example, setting a breakpoint has a synchronous response, but the corresponding breakpoint-hit is an asynchronous event that gets sent from NativeJDB to the IDE. In principle, NativeJDB can be used with any IDE that has a JDWP implementation, but it has mostly been tested and works with IntelliJ. It also does not require any changes or extensions to existing IDEs and works with the *community as well as the commercial version* of IntelliJ.
```

File: _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc (hunk 1 above)
Comment (vazirim) [reply _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc:28]: yes, native compilation with graalvm works only on Linux at the moment.

File: _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+
+[#conclusion]
+== Conclusion
+
+
+Through this blog, we have showcased a new debugging tool called NativeJDB that allows users to remotely attach and debug a natively compiled Java code.
+
+
+We look forward to getting feedback from you!
```
Comment (maxandersen) [inline _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc:135]: I would add a bit more text here explaining it is a working prototype that enables use of GDB in existing IDE tooling and that we are looking for feedback/suggestions. Maybe even mention idea of Microsoft debug protocol as future potential target ?

Comment (vazirim) [reply _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc:135]: Sounds good for adding more detail.

The Microsoft Debug Adapter protocol would be useful if we were extending existing IDEs, right? We are at a different point in the design space, where we want to allow existing IDEs to do remote debugging and attach to NativeJDB. So I am inclined not to include that. What do you think?

Comment (maxandersen) [reply _posts/2022-06-17-nativejdb-debugger-for-native-images.adoc]: sure, works for me

**Comment by aavarghese** [comment]
Looking at the preview, I can't find the names of the authors: Ansu Varghese and Mandana Vaziri. Could you please help?

image

**Comment by maxandersen** [comment]
the author tag is a singular author - we don't directly support for co-authors. sorry.

you'll have to pick one primary and then you can mention co-authors in the first text.

**Comment by aavarghese** [comment]
> the author tag is a singular author - we don't directly support for co-authors. sorry.
> 
> you'll have to pick one primary and then you can mention co-authors in the first text.

Ah, gotcha. Ok I'll make the change soon-ish. Thank you

**Comment by aavarghese** [comment]
@maxandersen done with the author changes...

**Comment by aavarghese** [comment]
@maxandersen ok to merge this soon?

**Review by maxandersen — APPROVED** [review]

**Comment by maxandersen** [comment]
done - thnks!

**Pull request merged**